### ISSUE TITLE
Transmission of models & textures in MP

### ISSUE THREAD

**Issue by ghost** [Nightfall/Instrumentality#7]
It's no good if we're all the same character...
Here's the general idea:
1. The custom renderer doesn't run unless the server says a player chose a model, or it's the local player and the user has chosen a model
2. When a player chooses a model they upload the binary blobs to the server. (Not performed if server has files-have a way to check)
3. Player then sends a manifest of which blob is the model, and which blobs map to what files
4. Server then deletes any unused files that player owns (OPTIONAL)
5. Server relays model to other players
6. When a player receives a model they then use that model for rendering the player associated with it

The server should be able to opt-out from transmitting/receiving files.
Since the client should have a list of models available anyway, the server must still pass the manifests - the clients will then search their local model directories for a PMX file that matches and try to use that.

**Comment by Victorious3** [comment]
Yeah, if this eventually gets a server side backend, definitely. That's what ModelCache is supposed to do. I'm not too sure about uploading models, because well... I'm always a bit paranoid if it comes to uploading something from the client -> server, but well. It would work that way. 

Not sure how big these files are though. I would throw it at a worker thread and then slowly load the models afterwards. We need a way of identifying a model.

**Comment by skyem123** [comment]
Is there any danger to loading the model files onto computers?

**Comment by Victorious3** [comment]
@skyem123 Well, it could be a virus ^^ It won't get run, but still. Maybe there is a way to validate the file...

**Comment by ghost** [comment]
To ID the model, we just get the hash of the uncompressed-but-packed file(something like a TAR). That's why we need the way the packing occurs to be consistent - so the same data creates the same packed file, which has the same hash.
As for the virus problem: there is no virus problem. Nobody is going to execute these files, putting executable files in would be detected as "unused files", and using them as a texture would cause a load error.

**Comment by skyem123** [comment]
so it'll be the same security issues as downloading a texture pack from the server?

**Comment by ghost** [comment]
skyem123: What security issues
The client isn't executing it...!

**Comment by skyem123** [comment]
Maybe there could be a way to share the animation files, but download the models from the official sources?
